File: packages/detector-core/aiprojectdetector/reporting.py
"""Build visualization payloads and recommendations from analysis artifacts."""

from __future__ import annotations

import os
from collections import Counter

from .ingestion.models import Scan
from .models import (
    AIScore,
    Classification,
    EntityAnalysis,
    VisualizationData,
)
from .utils.text import ngrams, tokenize


def build_visualizations(
    scan: Scan,
    files: list[EntityAnalysis],
    folders: list[EntityAnalysis],
    overall: AIScore,
    signal_means: dict[str, dict[str, float]],
) -> VisualizationData:
    viz = VisualizationData()

    viz.file_heatmap = [
        {
            "path": f.path,
            "ai_probability": f.score.ai_probability,
            "confidence": f.score.confidence,
            "classification": f.score.classification.value,
            "loc": f.loc,
            "language": f.language,
        }
        for f in files
    ]

    viz.folder_heatmap = [
        {
            "path": f.path,
            "ai_probability": f.score.ai_probability,
            "confidence": f.score.confidence,
            "classification": f.score.classification.value,
            "files": int(f.score.reasons[0].split()[0]) if f.score.reasons else 0,
        }
        for f in folders
    ]

    pie = Counter(f.score.classification.value for f in files)
    viz.classification_pie = [{"label": k, "value": v} for k, v in pie.items()]

    viz.commit_timeline = [
        {
            "sha": c.sha[:10],
            "date": c.timestamp.isoformat() if c.timestamp else None,
            "insertions": c.insertions,
            "deletions": c.deletions,
            "files_changed": c.files_changed,
            "message": c.message[:80],
        }
        for c in scan.commits
    ]

    viz.contributor_activity = [
        {
            "name": c.name,
            "commits": c.commits,
            "insertions": c.insertions,
            "deletions": c.deletions,
            "first_commit": c.first_commit.isoformat() if c.first_commit else None,
            "last_commit": c.last_commit.isoformat() if c.last_commit else None,
        }
        for c in scan.contributors
    ]

    viz.signal_breakdown = [
        {
            "signal": name,
            "mean_score": round(stats["mean_score"], 4),
            "count": int(stats["count"]),
            "mean_confidence": round(stats["mean_confidence"], 4),
        }
        for name, stats in sorted(
            signal_means.items(), key=lambda kv: abs(kv[1]["mean_score"] - 0.5), reverse=True
        )
    ]

    viz.similarity_matrix = _similarity_matrix(scan)
    viz.dependency_graph = _dependency_graph(scan)
    return viz


def _similarity_matrix(scan: Scan, top_n: int = 15) -> dict:
    """Jaccard similarity of 4-gram token sets for the largest analyzable files."""
    files = sorted(scan.analyzable(), key=lambda f: f.loc, reverse=True)[:top_n]
    if len(files) < 2:
        return {}
    grams = []
    labels = []
    for f in files:
        toks = tokenize(f.source)
        grams.append(set(ngrams(toks, 4)))
        labels.append(f.rel_path)
    matrix = []
    for a in grams:
        row = []
        for b in grams:
            if not a or not b:
                row.append(0.0)
                continue
            inter = len(a & b)
            union = len(a | b)
            row.append(round(inter / union, 3) if union else 0.0)
        matrix.append(row)
    return {"labels": labels, "matrix": matrix}


def _dependency_graph(scan: Scan, max_nodes: int = 120) -> dict:
    """Lightweight intra-repo import graph for Python and JS/TS files."""
    by_module: dict[str, str] = {}
    for f in scan.analyzable():
        stem = f.rel_path.rsplit(".", 1)[0]
        by_module[stem.replace("/", ".")] = f.rel_path
        by_module[os.path.basename(stem)] = f.rel_path

    nodes = []
    edges = []
    seen = set()
    for f in scan.analyzable()[:max_nodes]:
        nodes.append({"id": f.rel_path, "language": f.language, "loc": f.loc})
        seen.add(f.rel_path)
        for imp in _imports(f.source, f.language):
            target = by_module.get(imp) or by_module.get(imp.replace("/", "."))
            if target and target != f.rel_path:
                edges.append({"source": f.rel_path, "target": target})
    edges = [e for e in edges if e["source"] in seen and e["target"] in seen]
    return {"nodes": nodes, "edges": edges}


def _imports(source: str, language: str) -> list[str]:
    import re

    out: list[str] = []
    if language == "python":
        for m in re.finditer(r"^\s*(?:from\s+([\w.]+)\s+import|import\s+([\w.]+))", source, re.M):
            out.append((m.group(1) or m.group(2)).split(".")[0])
            if m.group(1):
                out.append(m.group(1))
    elif language in {"javascript", "typescript"}:
        for m in re.finditer(r"""(?:import[^'"]*from\s*|require\(\s*)['"]([^'"]+)['"]""", source):
            mod = m.group(1)
            if mod.startswith("."):
                out.append(os.path.normpath(mod).replace("\\", "/").lstrip("./"))
    return out


def build_recommendations(
    overall: AIScore,
    scan: Scan,
    files: list[EntityAnalysis],
) -> list[str]:
    recs: list[str] = [
        "Treat every score here as a probabilistic estimate, never as proof of AI authorship.",
    ]
    flagged = [f for f in files if f.score.ai_probability >= 0.62 and f.score.confidence >= 0.3]
    if flagged:
        top = sorted(flagged, key=lambda f: f.score.risk_score, reverse=True)[:5]
        recs.append(
            "Manually review the highest-risk files before drawing conclusions: "
            + ", ".join(f.path for f in top if f.path)
        )
    if not scan.git_available:
        recs.append(
            "No git history was available. Commit-pattern signals were skipped; "
            "providing the full repository (not just a snapshot) improves accuracy."
        )
    elif len(scan.commits) < 5:
        recs.append(
            "Very little commit history was available, which limits timeline-based signals."
        )
    if overall.confidence < 0.3:
        recs.append(
            "Overall confidence is low. The evidence is insufficient for a strong conclusion; "
            "gather more code or context before acting."
        )
    if overall.classification in {
        Classification.likely_ai_generated,
        Classification.likely_ai_assisted,
    }:
        recs.append(
            "If this assessment affects a person (e.g. academic or hiring decisions), "
            "corroborate with non-automated methods and give them a chance to respond."
        )
    return recs

File: packages/detector-core-ts/src/reporting.ts
/** Build visualization payloads and recommendations from analysis artifacts. */

import {
  makeVisualizationData,
  type AIScore,
  type Classification,
  type EntityAnalysis,
  type VisualizationData,
} from "./models.js";
import { scanAnalyzable, type Scan } from "./ingestion/models.js";
import { fileLoc } from "./ingestion/models.js";
import { ngrams, round, tokenize } from "./utils/text.js";
import { basename } from "./parsing/languages.js";

export interface SignalMeanStats {
  mean_score: number;
  mean_confidence: number;
  count: number;
}

export function buildVisualizations(
  scan: Scan,
  files: EntityAnalysis[],
  folders: EntityAnalysis[],
  _overall: AIScore,
  signalMeans: Record<string, SignalMeanStats>,
): VisualizationData {
  const viz = makeVisualizationData();

  viz.file_heatmap = files.map((f) => ({
    path: f.path,
    ai_probability: f.score.ai_probability,
    confidence: f.score.confidence,
    classification: f.score.classification,
    loc: f.loc,
    language: f.language,
  }));

  viz.folder_heatmap = folders.map((f) => ({
    path: f.path,
    ai_probability: f.score.ai_probability,
    confidence: f.score.confidence,
    classification: f.score.classification,
    files: f.score.reasons.length ? parseInt(f.score.reasons[0]!.split(/\s+/)[0]!, 10) || 0 : 0,
  }));

  const pie = new Map<string, number>();
  for (const f of files) pie.set(f.score.classification, (pie.get(f.score.classification) ?? 0) + 1);
  viz.classification_pie = Array.from(pie.entries()).map(([label, value]) => ({ label, value }));

  viz.commit_timeline = scan.commits.map((c) => ({
    sha: c.sha.slice(0, 10),
    date: c.timestamp,
    insertions: c.insertions,
    deletions: c.deletions,
    files_changed: c.files_changed,
    message: c.message.slice(0, 80),
  }));

  viz.contributor_activity = scan.contributors.map((c) => ({
    name: c.name,
    commits: c.commits,
    insertions: c.insertions,
    deletions: c.deletions,
    first_commit: c.first_commit,
    last_commit: c.last_commit,
  }));

  viz.signal_breakdown = Object.entries(signalMeans)
    .sort((a, b) => Math.abs(b[1].mean_score - 0.5) - Math.abs(a[1].mean_score - 0.5))
    .map(([name, stats]) => ({
      signal: name,
      mean_score: round(stats.mean_score, 4),
      count: Math.trunc(stats.count),
      mean_confidence: round(stats.mean_confidence, 4),
    }));

  viz.similarity_matrix = similarityMatrix(scan);
  viz.dependency_graph = dependencyGraph(scan);
  return viz;
}

/** Jaccard similarity of 4-gram token sets for the largest analyzable files. */
function similarityMatrix(scan: Scan, topN = 15): Record<string, unknown> {
  const files = scanAnalyzable(scan)
    .slice()
    .sort((a, b) => fileLoc(b) - fileLoc(a))
    .slice(0, topN);
  if (files.length < 2) return {};
  const grams: Set<string>[] = [];
  const labels: string[] = [];
  for (const f of files) {
    grams.push(new Set(ngrams(tokenize(f.source), 4)));
    labels.push(f.rel_path);
  }
  const matrix: number[][] = [];
  for (const a of grams) {
    const row: number[] = [];
    for (const b of grams) {
      if (a.size === 0 || b.size === 0) {
        row.push(0.0);
        continue;
      }
      let inter = 0;
      for (const g of a) if (b.has(g)) inter++;
      const union = a.size + b.size - inter;
      row.push(union ? round(inter / union, 3) : 0.0);
    }
    matrix.push(row);
  }
  return { labels, matrix };
}

/** Lightweight intra-repo import graph for Python and JS/TS files. */
function dependencyGraph(scan: Scan, maxNodes = 120): Record<string, unknown> {
  const byModule = new Map<string, string>();
  for (const f of scanAnalyzable(scan)) {
    const stem = f.rel_path.includes(".") ? f.rel_path.slice(0, f.rel_path.lastIndexOf(".")) : f.rel_path;
    byModule.set(stem.replace(/\//g, "."), f.rel_path);
    byModule.set(basename(stem), f.rel_path);
  }

  const nodes: Array<Record<string, unknown>> = [];
  const edges: Array<{ source: string; target: string }> = [];
  const seen = new Set<string>();
  for (const f of scanAnalyzable(scan).slice(0, maxNodes)) {
    nodes.push({ id: f.rel_path, language: f.language, loc: fileLoc(f) });
    seen.add(f.rel_path);
    for (const imp of imports(f.source, f.language)) {
      const target = byModule.get(imp) ?? byModule.get(imp.replace(/\//g, "."));
      if (target && target !== f.rel_path) edges.push({ source: f.rel_path, target });
    }
  }
  const filtered = edges.filter((e) => seen.has(e.source) && seen.has(e.target));
  return { nodes, edges: filtered };
}

function normalizePath(p: string): string {
  // Minimal browser-safe os.path.normpath for relative POSIX paths.
  const parts = p.replace(/\\/g, "/").split("/");
  const out: string[] = [];
  for (const part of parts) {
    if (part === "" || part === ".") continue;
    if (part === "..") {
      if (out.length && out[out.length - 1] !== "..") out.pop();
      else out.push("..");
    } else out.push(part);
  }
  return out.join("/");
}

function imports(source: string, language: string): string[] {
  const out: string[] = [];
  if (language === "python") {
    const re = /^\s*(?:from\s+([\w.]+)\s+import|import\s+([\w.]+))/gm;
    for (const m of source.matchAll(re)) {
      const full = (m[1] ?? m[2])!;
      out.push(full.split(".")[0]!);
      if (m[1]) out.push(m[1]);
    }
  } else if (language === "javascript" || language === "typescript") {
    const re = /(?:import[^'"]*from\s*|require\(\s*)['"]([^'"]+)['"]/g;
    for (const m of source.matchAll(re)) {
      const mod = m[1]!;
      if (mod.startsWith(".")) out.push(normalizePath(mod).replace(/^\.\//, ""));
    }
  }
  return out;
}

export function buildRecommendations(
  overall: AIScore,
  scan: Scan,
  files: EntityAnalysis[],
): string[] {
  const recs: string[] = [
    "Treat every score here as a probabilistic estimate, never as proof of AI authorship.",
  ];
  const flagged = files.filter((f) => f.score.ai_probability >= 0.62 && f.score.confidence >= 0.3);
  if (flagged.length) {
    const top = flagged
      .slice()
      .sort((a, b) => b.score.risk_score - a.score.risk_score)
      .slice(0, 5);
    recs.push(
      "Manually review the highest-risk files before drawing conclusions: " +
        top
          .map((f) => f.path)
          .filter((p): p is string => !!p)
          .join(", "),
    );
  }
  if (!scan.git_available) {
    recs.push(
      "No git history was available. Commit-pattern signals were skipped; " +
        "providing the full repository (not just a snapshot) improves accuracy.",
    );
  } else if (scan.commits.length < 5) {
    recs.push(
      "Very little commit history was available, which limits timeline-based signals.",
    );
  }
  if (overall.confidence < 0.3) {
    recs.push(
      "Overall confidence is low. The evidence is insufficient for a strong conclusion; " +
        "gather more code or context before acting.",
    );
  }
  const aiBuckets: Classification[] = ["likely_ai_generated", "likely_ai_assisted"];
  if (aiBuckets.includes(overall.classification)) {
    recs.push(
      "If this assessment affects a person (e.g. academic or hiring decisions), " +
        "corroborate with non-automated methods and give them a chance to respond.",
    );
  }
  return recs;
}
